In [2]:
# Importando bibliotecas
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix

# Carregando os dados
df = pd.read_csv('creditcard_2023.csv')

# Separando recursos e rótulos
X = df.drop(['Class'], axis=1)  # Recursos
y = df['Class']  # Rótulo

# Dividindo os dados em conjuntos de treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

X_train = X_train.values
X_test = X_test.values


# Construindo e treinando o modelo KNN
knn = KNeighborsClassifier(n_neighbors=5)  # Você pode ajustar o número de vizinhos
knn.fit(X_train, y_train)

# Fazendo previsões e avaliando o modelo
y_pred = knn.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))


[[85139    10]
 [  143 85297]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85149
           1       1.00      1.00      1.00     85440

    accuracy                           1.00    170589
   macro avg       1.00      1.00      1.00    170589
weighted avg       1.00      1.00      1.00    170589



Os resultados que você forneceu incluem uma matriz de confusão e um relatório de classificação do seu modelo K-Nearest Neighbors (KNN) usado para a detecção de fraude em cartões de crédito. Vamos interpretá-los:

### Matriz de Confusão
A matriz de confusão é:

```
[[85139    10]
 [  143 85297]]
```

- **Verdadeiros Negativos (TN)**: 85,139 - O modelo previu corretamente a classe negativa (não-fraude) 85,139 vezes.
- **Falsos Positivos (FP)**: 10 - O modelo previu erroneamente a classe positiva (fraude) 10 vezes quando era realmente negativa.
- **Falsos Negativos (FN)**: 143 - O modelo previu erroneamente a classe negativa 143 vezes quando era realmente positiva.
- **Verdadeiros Positivos (TP)**: 85,297 - O modelo previu corretamente a classe positiva 85,297 vezes.

### Relatório de Classificação
O relatório de classificação mostra as seguintes métricas:

- **Precision** (Precisão): A precisão para ambas as classes é de 1,00, o que indica que, quando o modelo prevê uma classe, ele está correto na grande maioria das vezes.

- **Recall** (Revocação ou Sensibilidade): O recall também é de 1,00 para ambas as classes, significando que o modelo foi capaz de identificar corretamente a grande maioria das instâncias de ambas as classes.

- **F1-Score**: O F1-Score é a média harmônica de precisão e recall. Um F1-Score de 1,00 para ambas as classes indica um equilíbrio perfeito entre precisão e recall.

- **Support**: O número de ocorrências reais de cada classe no conjunto de dados especificado. Para a classe 0, são 85,149 instâncias, e para a classe 1, são 85,440 instâncias.

- **Accuracy** (Acurácia): A proporção de previsões corretas (tanto positivas quanto negativas) em relação ao total de previsões. A acurácia do modelo é extremamente alta.

### Interpretação Geral
O modelo KNN demonstrou um desempenho excepcional, com uma acurácia muito alta, e altas pontuações de precisão e recall para ambas as classes. Isso sugere que o modelo é muito eficaz na detecção de transações fraudulentas e não fraudulentas.

No entanto, é importante notar que, mesmo com um desempenho tão alto, há um pequeno número de falsos positivos e falsos negativos. Isso é especialmente significativo em aplicações como a detecção de fraude em cartões de crédito, onde tanto a identificação de transações legítimas (minimizando os falsos positivos) quanto a detecção de transações fraudulentas (minimizando os falsos negativos) são cruciais. 

Finalmente, como sempre, é importante estar atento ao risco de overfitting e garantir que o conjunto de teste seja representativo da população geral que o modelo irá encontrar no mundo real.